contracts/token/ERC20/extensions/draft-ERC20Votes.sol

```diff
@@ -0,0 +1,169 @@
+ * the base gas cost of transfers.
+ *
+ * _Available since v4.2._
+ */
+abstract contract ERC20Votes is IERC20Votes, ERC20Permit {
+    bytes32 private constant _DELEGATION_TYPEHASH = keccak256("Delegation(address delegatee,uint256 nonce,uint256 expiry)");
+
+    mapping (address => address) private _delegates;
+    mapping (address => Checkpoint[]) private _checkpoints;
+
+    function checkpoints(address account, uint32 pos) external view virtual override returns (Checkpoint memory) {
+        return _checkpoints[account][pos];
+    }
+
+    function numCheckpoints(address account) external view virtual override returns (uint32) {
+        return uint32(_checkpoints[account].length);
+    }
+
+    /**
+    * @dev Get the address `account` is currently delegating to.
+    */
+    function delegates(address account) public view virtual override returns (address) {
+        return _delegates[account];
+    }
+
+    /**
+     * @dev Example: This enables autodelegation, makes each transfer more expensive but doesn't require user to
+     * delegate to themselves. Can be usefull for tokens useds exclusivelly for governance, such as voting wrappers of
+     * pre-existing ERC20.
+     */
+    // function delegates(address account) public view override returns (address) {
+    //     address delegatee = _delegates[account];
+    //     return delegatee == address(0) ? account : delegatee;
+    // }
+
+    /**
+     * @notice Gets the current votes balance for `account`
+     * @param account The address to get votes balance
+     * @return The number of current votes for `account`
+     */
+    function getCurrentVotes(address account) external view override returns (uint256) {
+        uint256 pos = _checkpoints[account].length;
+        return pos == 0 ? 0 : _checkpoints[account][pos - 1].votes;
+    }
+
+    /**
+     * @notice Determine the prior number of votes for an account as of a block number
+     * @dev Block number must be a finalized block or else this function will revert to prevent misinformation.
+     * @param account The address of the account to check
+     * @param blockNumber The block number to get the vote balance at
+     * @return The number of votes the account had as of the given block
+     */
+    function getPriorVotes(address account, uint256 blockNumber) external view override returns (uint256) {
+        require(blockNumber < block.number, "ERC20Votes::getPriorVotes: not yet determined");
+
+        Checkpoint[] storage ckpts = _checkpoints[account];
+
+        uint256 high = ckpts.length;
+        uint256 low = 0;
+        while (low < high) {
+            uint256 mid = Math.average(low, high);
+            if (ckpts[mid].fromBlock > blockNumber) {
+                high = mid;
+            } else {
+                low = mid + 1;
+            }
+        }
+
+        return low == 0 ? 0 : ckpts[low - 1].votes;
```

Comment:
This doesn't look right to me. What happens when you query for the first checkpointed block? I think this returns zero.

Line 84 should be `low = mid` and here we should return `ckpts[low]` (unless...).

`high` is the exclusive upper bound of the range that contains the right checkpoint, and `low` is an inclusive lower bound. 

Please confirm. Tests for this scenario would be good.

Reply:
We have this test, which I believe covers it:

      it('returns the number of checkpoints for a delegate', async function () {
        await this.token.transfer(recipient, '100', { from: holder }); //give an account a few tokens for readability
        expect(await this.token.numCheckpoints(other1)).to.be.bignumber.equal('0');

        const t1 = await this.token.delegate(other1, { from: recipient });
        expect(await this.token.numCheckpoints(other1)).to.be.bignumber.equal('1');

        const t2 = await this.token.transfer(other2, 10, { from: recipient });
        expect(await this.token.numCheckpoints(other1)).to.be.bignumber.equal('2');

        const t3 = await this.token.transfer(other2, 10, { from: recipient });
        expect(await this.token.numCheckpoints(other1)).to.be.bignumber.equal('3');

        const t4 = await this.token.transfer(recipient, 20, { from: holder });
        expect(await this.token.numCheckpoints(other1)).to.be.bignumber.equal('4');

        expect(await this.token.checkpoints(other1, 0)).to.be.deep.equal([ t1.receipt.blockNumber.toString(), '100' ]);
        expect(await this.token.checkpoints(other1, 1)).to.be.deep.equal([ t2.receipt.blockNumber.toString(), '90' ]);
        expect(await this.token.checkpoints(other1, 2)).to.be.deep.equal([ t3.receipt.blockNumber.toString(), '80' ]);
        expect(await this.token.checkpoints(other1, 3)).to.be.deep.equal([ t4.receipt.blockNumber.toString(), '100' ]);
      });```

Reply:
walkthrough, assuming `length = 4` and `ckpts[0] = blocknumber` 

- low = 0, high = 4, mid = 2 → update high
- low = 0, high = 2, mid = 1 → update high
- low = 0, high = 1, mid = 0 → update low
- low = 1, high = 1, loop stops
- we read ckpts[low-1] → good!

Reply:
For low to be 0 after the loop end, you'd need low = high = 0, meaning that there are either no checkpoints at all, or for all checkpoints, including ckpts[0], `ckpts[mid].fromBlock > blockNumber` ... this means that all checkpoints are older then the request, and thus returning 0 is the right thing to do

Reply:
Also, if we do low = mid, we can get stuck in infinite loops:
low = 0, high = 1, mid = 0 ... doing low = mid will loop forever